File: examples/java-jfr/rideshare/src/main/java/org/example/rideshare/OrderService.java
package org.example.rideshare;

import io.pyroscope.labels.LabelsSet;
import io.pyroscope.labels.Pyroscope;
import org.springframework.stereotype.Service;

import java.time.Duration;
import java.time.Instant;
import java.time.ZoneOffset;
import java.time.temporal.ChronoUnit;
import java.util.concurrent.atomic.AtomicLong;

@Service
public class OrderService {

    public static final Duration OP_DURATION = Duration.of(200, ChronoUnit.MILLIS);

    public synchronized void findNearestVehicle(int searchRadius, String vehicle) {
        Pyroscope.LabelsWrapper.run(new LabelsSet("vehicle", vehicle), () -> {
            AtomicLong i = new AtomicLong();
            Instant end = Instant.now()
                    .plus(OP_DURATION.multipliedBy(searchRadius));
            while (Instant.now().compareTo(end) <= 0) {
                i.incrementAndGet();
            }

            if (vehicle.equals("car")) {
                checkDriverAvailability(searchRadius);
            }
        });
    }

    private void checkDriverAvailability(int searchRadius) {
        AtomicLong i = new AtomicLong();
        Instant end = Instant.now()
                .plus(OP_DURATION.multipliedBy(searchRadius));
        while (Instant.now().compareTo(end) <= 0) {
            i.incrementAndGet();
        }
        // Every other minute this will artificially create make requests in eu-north region slow
        // this is just for demonstration purposes to show how performance impacts show up in the
        // flamegraph
        boolean force_mutex_lock = Instant.now().atZone(ZoneOffset.UTC).getMinute() % 2 == 0;
        if (System.getenv("REGION").equals("eu-north") && force_mutex_lock) {
            mutexLock(searchRadius);
        }
    }

    private void mutexLock(int searchRadius) {
        AtomicLong i = new AtomicLong();
        Instant end = Instant.now()
                .plus(OP_DURATION.multipliedBy(30L * searchRadius));
        while (Instant.now().compareTo(end) <= 0) {
            i.incrementAndGet();
        }
    }

}

File: examples/dotnet/rideshare/example/OrderService.cs
using System;

namespace Example;

internal class OrderService
{
    public void FindNearestVehicle(long searchRadius, string vehicle)
    {
        lock (_lock)
        {
            var labels = Pyroscope.LabelSet.Empty.BuildUpon()
                .Add("vehicle", vehicle)
                .Build();
            Pyroscope.LabelsWrapper.Do(labels, () =>
            {
                for (long i = 0; i < searchRadius * 1000000000; i++)
                {
                }

                if (vehicle.Equals("car"))
                {
                    CheckDriverAvailability(labels, searchRadius);
                }
            });
        }
    }

    private readonly object _lock = new();

    private static void CheckDriverAvailability(Pyroscope.LabelSet ctx, long searchRadius)
    {
        var region = System.Environment.GetEnvironmentVariable("REGION") ?? "unknown_region";
        ctx = ctx.BuildUpon()
            .Add("driver_region", region)
            .Build();
        Pyroscope.LabelsWrapper.Do(ctx, () =>
        {
            for (long i = 0; i < searchRadius * 1000000000; i++)
            {
            }

            var now = DateTime.Now.Minute % 2 == 0;
            var forceMutexLock = DateTime.Now.Minute % 2 == 0;
            if ("eu-north".Equals(region) && forceMutexLock)
            {
                MutexLock(searchRadius);
            }
        });
    }

    private static void MutexLock(long searchRadius)
    {
        for (long i = 0; i < 30 * searchRadius * 1000000000; i++)
        {
        }
    }
}